text/0000-api-evolution.md

```diff
@@ -0,0 +1,748 @@
+
+### Principles of the policy
+
+The basic design of the policy is that **minor changes should require at most a
+few local *annotations* to the code you are developing, and in principle no
+changes to your dependencies.**
+
+In more detail:
+
+* Minor changes should require at most minor amounts of work upon upgrade. For
+  example, changes that may require occasional type annotations or use of UFCS
+  to disambiguate are not automatically "major" changes. (But in such cases, one
+  must evaluate how widespread these "minor" changes are).
+
+* In principle, it should be possible to produce a version of the code for any
+  dependencies that *will not break* when upgrading to a new minor
+  revision. This goes hand-in-hand with the above bullet; as we will see, it's
+  possible to save a fully "elaborated" version of upstream code that does not
+  require any disambiguation. The "in principle" refers to the fact that getting
+  there may require some additional tooling or language support, which this RFC
+  outlines.
+
+That means that any breakage in a minor release must be very "shallow": it must
+always be possible to locally fix the problem through some kind of
+disambiguation *that could have been done in advance* (by using more explicit
+forms) or other annotation (like disabling a lint). It means that minor changes
+can never leave you in a state that requires breaking changes to your own code.
+
+**Although this general policy allows some (very limited) breakage in minor
+releases, it is not a license to make these changes blindly**. The breakage that
+this RFC permits, aside from being very simple to fix, is also unlikely to occur
+often in practice. The RFC will discuss measures that should be employed in the
+standard library to ensure that even these minor forms of breakage do not cause
+widespread pain in the ecosystem.
+
+### Scope of the policy
+
+The policy laid out by this RFC applies to *stable*, *public* APIs in the
+standard library. Eventually, stability attributes will be usable in external
+libraries as well (this will require some design work), but for now public APIs
+in external crates should be understood as de facto stable after the library
+reaches 1.0.0 (per semver).
+
+## Policy by language feature
+
+Most of the policy is simplest to lay out with reference to specific language
+features and the way that APIs using them can, and cannot, evolve in a minor
+release.
+
+**Breaking changes are assumed to be major changes unless otherwise stated**.
+The RFC covers many, but not all breaking changes that are major; it covers
+*all* breaking changes that are considered minor.
+
+### Crates
+
+#### Major change: introducing `#[feature]` for the first time.
+
+Changing a crate from working on stable Rust to *requiring* a nightly is
+considered a breaking change. Crate authors should consider using Cargo
+"features" for their crate to make such use opt-in.
+
+#### Minor change: adding/removing crate dependencies.
```

Comment:
A somewhat related concept would be changing the major version of a dependency. It seems like that could be either a minor or major change depending on if the crate is visible in your crate's public API.